feat: add snapshotting to accurately calculate claimable amounts at any time

ramenforbreakfast opened the pull request:
Since we need to be able to distinguish vested amounts (but already claimed) versus those that have not been claimed, and know these amounts at any given time in the past for snapshot voting, we need to add snapshots and rework our `getClaimableAmount` and `getClaimedAmount` functions.

This PR is quite large, but these features really can't be detached since they are so tightly connected, here is a summary of these changes:
- Unfortunately, checkpoints are not an OZ module that I can import, however, I reused a lot of code present in https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v4.8.1/contracts/token/ERC20/extensions/ERC20Votes.sol
- This contract uses the same `Checkpoint` system, with some slight modifications, mainly to accomodate `timestamp`, rather than `blockNumber`. The choice to simplify the structure and simply have two `uint256`s, rather than tightly packing to save bits, is mostly to avoid having to perform casting and `uint32` being questionable for a timestamp.
- `getClaimedAmount` and `getClaimableAmount` now have 2 versions, one that gets the current value, and one that allows a specifiable timestamp. These both call `internal` versions to reduce code duplication.
- `_getClaimedAmount` uses a binary search algorithm to find the `Checkpoint` that is closest to the specified timestamp, this was taken from the OZ code linked to above.
- Note: You will see in the OZ code that a internal function called `_unsafeAccess` is used, which is an optimization for accessing array entries. This function was causing issues when I tried to reuse it for our purposes and so I decided to simply access the entries normally.

XavierEkkel commented on apps/smart-contracts/token/test/Vesting.test.ts:
```diff
@@ -503,19 +509,97 @@ describe('Vesting', () => {
+       * Ensure invariants are not violated across multiple claims,
+       * for both retroactive and future timestamps.
+       */
+      for (let i = 0; i < numberOfClaims; i++) {
```
could these complicated loops be extracted out into a helper function?

ramenforbreakfast replied:
Resolving this since a follow up PR will be adding testcases that will not have for loops

XavierEkkel commented on apps/smart-contracts/token/test/Vesting.test.ts:
```diff
@@ -503,19 +509,97 @@ describe('Vesting', () => {
+        // eslint-disable-next-line no-restricted-syntax
+        for (let j = 0; j < numberOfClaims; j++) {
+          if (j <= i) {
+            const claimedAmountJustBefore = await vesting['getClaimedAmount(address,uint256)'](
```
```suggestion
            const previousClaimedAmount = await vesting['getClaimedAmount(address,uint256)'](
```

XavierEkkel commented on apps/smart-contracts/token/test/Vesting.test.ts:
```diff
@@ -503,19 +509,97 @@ describe('Vesting', () => {
+          expect(
+            await vesting['getClaimedAmount(address,uint256)'](user1.address, claimingTimes[j - 1])
+          ).eq(
+            await vesting['getClaimedAmount(address,uint256)'](user1.address, claimingTimes[j] - 1)
```
```suggestion
            await vesting['getClaimedAmount(address,uint256)'](user1.address, claimingTimes[j] - amountsAllocated[0])
```

ramenforbreakfast replied:
I think you are misunderstanding what this testcase is doing.
The `[j-1]` here, means _The last time a claim happened_
The `[j]-1` here, means _Right before the current claim_

The time in between these points is the time in between claims, and we are ensuring the value does not change in-between these times, which is what the test case states it is testing.

ramenforbreakfast replied:
Resolving this since a follow up PR will be adding testcases that will not have for loops

XavierEkkel commented on apps/smart-contracts/token/test/Vesting.test.ts:
```diff
@@ -503,19 +509,97 @@ describe('Vesting', () => {
-  describe('# getClaimableAmount', () => {
+  describe('# getClaimedAmount(address,uint256)', () => {
+    before(() => {
+      amountsAllocated = [parseEther('1')]
```
nit:
`amountsAllocated` as the array is only used once, whereas `amountsAllocated[0]` is used repeatedly

make `amountsAllocated[0]` a variable instead, and use `[amountsAllocated]` for the one time you need the array

ramenforbreakfast replied:
This is only being done because `amountsAllocated` is used by all other testcases (I'm not a fan of how this is done in this test suite, I didnt write it and didn't want to waste time refactoring this) and is already an array type, i cannot turn it into a standalone number type like you want.

XavierEkkel commented on apps/smart-contracts/token/test/Vesting.test.ts:
```diff
@@ -503,19 +509,97 @@ describe('Vesting', () => {
+      const amountToClaimEachTime = amountsAllocated[0].div(numberOfClaims)
+
+      /**
+       * Ensure invariants are not violated across multiple claims,
```
it's just a single invariant (the eq check) right? (not multiple)

ramenforbreakfast replied:
I just added this comment to explain why the testcases following all use this double for loop structure. It is so that we can check all relevant timestamps at each claim, for multiple claims.

ramenforbreakfast replied:
Yes!

XavierEkkel commented on apps/smart-contracts/token/test/Vesting.test.ts:
```diff
@@ -503,19 +509,97 @@ describe('Vesting', () => {
+            const claimedAmountJustBefore = await vesting['getClaimedAmount(address,uint256)'](
+              user1.address,
+              claimingTimes[j] - 1
+            )
```
```suggestion
            )
            
```

XavierEkkel commented on apps/smart-contracts/token/test/Vesting.test.ts:
```diff
@@ -503,19 +509,97 @@ describe('Vesting', () => {
+        await vesting.connect(user1).claim()
+        // eslint-disable-next-line no-restricted-syntax
+        for (let j = 0; j < numberOfClaims; j++) {
+          if (j <= i) {
```
can move this into the for loop condition check
and in general, strongly avoid nestings this deep

XavierEkkel commented on apps/smart-contracts/token/test/Vesting.test.ts:
```diff
@@ -503,19 +509,97 @@ describe('Vesting', () => {
+
+      /**
+       * Ensure invariants are not violated across multiple claims,
+       * for both retroactive and future timestamps.
```
generally not a good idea to test multiple types of scenarios in one test as if it fails it's hard to know why. ideally separate into separate tests w/ separate helpers. this may also make it easier to read+understand

ramenforbreakfast replied:
Im not sure if there is a misunderstanding, but this isn't "multiple scenarios". All this test does is assert that the claimed amount is added to the amount before claiming, when a claim happens. Everytime a claim happens, we are asserting this condition holds for all timestamps in the future, including the current timestamp.

ramenforbreakfast replied:
Resolving this since a follow up PR will be adding testcases that will not have for loops

ramenforbreakfast commented on apps/smart-contracts/token/test/Vesting.test.ts:
```diff
@@ -453,20 +455,20 @@ describe('Vesting', () => {
     it('transfers multiple times if still vesting', async () => {
       await mockERC20Token.transfer(vesting.address, amountsAllocated[0])
-      const numberOfWithdrawals = 2
+      const numberOfWithdrawals = 5
```
I felt 2 claims was not sufficient enough to demonstrate a condition holding across multiple timeframes

ramenforbreakfast replied:
Resolving since it has been decided to add additional simpler 2 round test cases in a follow up PR.

ramenforbreakfast commented on apps/smart-contracts/token/contracts/vesting/Vesting.sol:
```diff
@@ -162,6 +191,48 @@ contract Vesting is IVesting, Pausable, SafeOwnable, WithdrawERC20 {
+      high == 0 ? 0 : _recipientToClaimedAmount[recipient][high - 1].amount;
+  }
+
+  function _getClaimableAmount(address recipient, uint256 timestamp)
```
Not sure why we wouldn't just change it in this PR? It doesn't make the PR smaller because the function signatures need to be changed anyway due to the presence of 2 `getUnclaimedAmount` (previously `getClaimableAmount`) functions.

ramenforbreakfast commented on apps/smart-contracts/token/test/Vesting.test.ts:
```diff
@@ -503,19 +509,97 @@ describe('Vesting', () => {
+      await vesting.setAllocations([user1.address], amountsAllocated)
+      await mockERC20Token.transfer(vesting.address, amountsAllocated[0])
+    })
+
```
The end tests already exist, and kind of are overkill since this is dependent on claiming, not necessarily a specific time case. These end tests were bundled with the "never claimed" tests, and I have now just put them as separate "never claimed" tests.

ramenforbreakfast commented on apps/smart-contracts/token/test/Vesting.test.ts:
```diff
@@ -503,19 +509,97 @@ describe('Vesting', () => {
+      await vesting.setAllocations([user1.address], amountsAllocated)
+      await mockERC20Token.transfer(vesting.address, amountsAllocated[0])
+    })
+
```
I am instead going to create a test case to verify that a user does not affect another user's claimed amount, since it is inaccurate to say the function returns 0 if another user claims.

ramenforbreakfast commented:
This PR has been modified with the following major changes:
- `getClaimableAmount` has been changed to `getUnclaimedAmount`
- for loops for looping through different timestamps have been simplified to remove conditional logic that can just be moved into the loop condition.
- Added further commenting to for loop test cases
- Added a bunch of =0, start time, end time test cases for both `getClaimedAmount` and `getUnclaimedAmount`

Simplified versions of the more complex loop testcases will be submitted later.

XavierEkkel commented on apps/smart-contracts/token/test/Vesting.test.ts:
```diff
@@ -585,25 +726,210 @@ describe('Vesting', () => {
+      }
+    })
+
+    it('returns claimable amount at times in between a claim', async () => {
```
nit: change claimable -> unclaimed in these test case names

XavierEkkel reviewed (approved):
LGTM after nit
